src/agent.py

```diff
@@ -96,4 +101,112 @@ def _on_agent_relation_joined(self, event: RelationJoinedEvent) -> None:
+        if not container.can_connect():
+            event.defer()
+            return
+        if not event.unit or not all(
+            event.relation.data[event.unit].get(required_data)
+            for required_data in ("executors", "labels", "name")
+        ):
+            logger.warning("Relation data not ready yet. Deferring.")
+            event.defer()
+            return
+        agent_meta = jenkins.AgentMeta(
+            executors=event.relation.data[event.unit]["executors"],
+            labels=event.relation.data[event.unit]["labels"],
+            name=event.relation.data[event.unit]["name"],
+        )
+        try:
+            agent_meta.validate()
+        except jenkins.ValidationError as exc:
+            logger.error("Invalid agent relation data. %s", exc)
+            self.charm.unit.status = ops.BlockedStatus("Invalid agent relation data.")
+            return
+
+        self.charm.unit.status = ops.MaintenanceStatus("Adding agent node.")
+        credentials = jenkins.get_admin_credentials(container)
+        try:
+            jenkins.add_agent_node(
+                agent_meta=agent_meta,
+                credentials=credentials,
+            )
+            secret = jenkins.get_node_secret(credentials=credentials, node_name=agent_meta.name)
+        except jenkins.JenkinsError as exc:
+            self.charm.unit.status = ops.BlockedStatus(f"Jenkins API exception. {exc=!r}")
+            return
+
+        # This is to avoid the None type.
+        assert (binding := self.model.get_binding("juju-info"))  # nosec
+        host = binding.network.bind_address
+        event.relation.data[self.model.unit].update(
+            {
+                "url": f"http://{host}:{jenkins.WEB_PORT}",
+                f"{agent_meta.name}_secret": secret,
+            }
+        )
+        self.charm.unit.status = ops.ActiveStatus()
+
+    def _on_deprecated_agent_relation_departed(self, event: ops.RelationDepartedEvent) -> None:
+        """Handle deprecated agent relation departed event.
+
+        Args:
+            event: The event fired when a unit in deprecated agent relation is departed.
+        """
+        # the event unit cannot be None.
+        container = self.charm.unit.get_container(self.state.jenkins_service_name)
+        if not container.can_connect():
+            return
+
+        # The relation data is removed before this particular hook runs, making the name set by the
+        # agent not available. Hence, we can try to infer the name of the unit.
+        # See discussion: https://github.com/canonical/operator/issues/888
+        # assert type since event unit cannot be None.
+        agent_name = jenkins.get_agent_name(typing.cast(ops.Unit, event.unit).name)
+        credentials = jenkins.get_admin_credentials(container)
+        self.charm.unit.status = ops.MaintenanceStatus("Removing agent node.")
+        try:
+            jenkins.remove_agent_node(agent_name=agent_name, credentials=credentials)
```

Comment:
Similar to above, can this function figure out the `agent_name` and credentials itself? In general, try and limit the number of interactions from the charm classes to the modules. If there are multiple function calls - it is probably worth creating a higher level function on the module that does what the multiple function calls do

Reply:
Refactored, thank you!